src/crimson/osd/ops_executer.cc

```diff
@@ -100,51 +100,219 @@ seastar::future<> OpsExecuter::do_op_call(OSDOp& osd_op)
     // this should only happen with a buggy client.
     throw std::invalid_argument("outside of PG bounds");
   }
+
   return backend.list_objects(lower_bound, limit).then(
-    [lower_bound, pg_end, nspace](auto objects, auto next) {
-      auto in_my_namespace = [&nspace](const hobject_t& o) {
+    [this, filter, nspace](auto objects, auto next) {
+      auto in_my_namespace = [&nspace](const hobject_t& obj) {
         using ceph::common::local_conf;
-        if (o.get_namespace() == local_conf()->osd_hit_set_namespace) {
+        if (obj.get_namespace() == local_conf()->osd_hit_set_namespace) {
           return false;
         } else if (nspace == librados::all_nspaces) {
           return true;
         } else {
-          return o.get_namespace() == nspace;
+          return obj.get_namespace() == nspace;
+        }
+      };
+      auto to_pglsed = [this, filter] (const hobject_t& obj) {
+        // this transformation looks costly. However, I don't have any
+        // reason to think PGLS* operations are critical for, let's say,
+        // general performance.
+        if (filter) {
+          return pgls_filter(*filter, obj);
+        } else {
+          return seastar::make_ready_future<bool, hobject_t>(true, obj);
         }
       };
+
+      auto range = objects | boost::adaptors::filtered(in_my_namespace)
+                           | boost::adaptors::transformed(to_pglsed);
+      logger().debug("do_pgnls_common: finishing the 1st stage of pgls");
+      return seastar::when_all_succeed(std::begin(range),
+                                       std::end(range)).then(
+        [next=std::move(next)] (auto items) mutable {
+          // the sole purpose of this chaining is to pass `next` to 2nd
+          // stage altogether with items
+          logger().debug("do_pgnls_common: 1st done");
+          return seastar::make_ready_future<decltype(items), decltype(next)>(
+            std::move(items), std::move(next));
+      });
+  }).then(
+    [pg_end, filter] (const std::vector<std::tuple<bool,
+                                hobject_t>>& items, auto next) {
+      auto is_matched = [] (const auto& item) {
+        return std::get<bool>(item);
+      };
+      auto to_entry = [] (const auto& item) {
+        const auto& obj = std::get<hobject_t>(item);
+        return librados::ListObjectImpl{
+          obj.get_namespace(), obj.oid.name, obj.get_key()
+        };
+      };
+
       pg_nls_response_t response;
-      boost::copy(objects |
-        boost::adaptors::filtered(in_my_namespace) |
-        boost::adaptors::transformed([](const hobject_t& o) {
-          return librados::ListObjectImpl{o.get_namespace(),
-                                          o.oid.name,
-                                          o.get_key()}; }),
-        std::back_inserter(response.entries));
+      boost::push_back(response.entries,
+                       items | boost::adaptors::filtered(is_matched)
```

Comment:
nit, i'd recommend put the operator at the end of line, so the reader is aware that the statement is not ended. for an exaggerated instance, 
```c++
int i = 0 + 1
          + 2
          + 3;
```

Reply:
Refactored. I hope that exceeding the 80th column is fine.